### ISSUE TITLE
Invalid username, password, security token; or user locked out.

### ISSUE THREAD

**Issue by drake-yesware** [springml/spark-salesforce#33]
I'm getting this error while trying to connect with password and token. 
I'm using `https://test.salesforce.com` as a login url.

I'm able to login to Salesforce Sandbox with these credentials, and I've tested the same credentials  with the security token on another Salesforce SDK and I'm able to connect without issue. 

I've tried this on both Spark 2.2.0 and 2.3.0, version 2.11:1.1.0 and 2.11:1.1.1 and I still can't connect. However, the first time I tried using this package several days ago, I was able to connect with the same code and same credentials. 

I'm not using IP white-listing on the sandbox instance. Could this issue be related to this package, or is it more likely to be something with Salesforce itself?

**Comment by springml** [comment]
Did you include api token as part of the password.

**Comment by drake-yesware** [comment]
@springml Yes. I've appended my security token to the end of the password.

**Comment by drake-yesware** [comment]
I've tried changing the login URL to the actual sandbox instance I'm logged into in the console. That doesn't work either. I've also reset my password / multiple times and my user is not locked out since I'm able to login both in the console as well as through the Node.js API. 

Any one have any ideas?

**Comment by springml** [comment]
Can you check by calling the api from a third party tool like SoapUI or Postman or even a simple Java/Python script?
Also what does your Spark read.df command look like?

**Comment by samuel-pt** [comment]
@drake-yesware - Do you have any stackTrace? If yes, please add it as a comment. It will be useful to identify the issue

**Comment by drake-yesware** [comment]
@samuel-pt  Here is the output from writing the dataframe above:

```
18/05/02 19:51:26 WARN salesforce.SFObjectWriter: SaveMode ErrorIfExists Not supported. Using 'insert' operation
18/05/02 19:51:26 ERROR util.SFConfig: Exception while creating connection
[LoginFault [ApiFault  exceptionCode='INVALID_LOGIN'
 exceptionMessage='Invalid username, password, security token; or user locked out.'
 extendedErrorDetails='{[0]}'
]
]

	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at com.sforce.ws.bind.TypeMapper.readSingle(TypeMapper.java:687)
	at com.sforce.ws.bind.TypeMapper.readObject(TypeMapper.java:562)
	at com.sforce.ws.transport.SoapConnection.parseDetail(SoapConnection.java:236)
	at com.sforce.ws.transport.SoapConnection.createException(SoapConnection.java:210)
	at com.sforce.ws.transport.SoapConnection.receive(SoapConnection.java:156)
	at com.sforce.ws.transport.SoapConnection.send(SoapConnection.java:99)
	at com.sforce.soap.partner.PartnerConnection.login(PartnerConnection.java:816)
	at com.sforce.soap.partner.PartnerConnection.<init>(PartnerConnection.java:426)
	at com.sforce.soap.partner.Connector.newConnection(Connector.java:27)
	at com.springml.salesforce.wave.util.SFConfig.createPartnerConnection(SFConfig.java:107)
	at com.springml.salesforce.wave.util.SFConfig.getPartnerConnection(SFConfig.java:90)
	at com.springml.salesforce.wave.impl.BulkAPIImpl.createJob(BulkAPIImpl.java:41)
	at com.springml.salesforce.wave.impl.BulkAPIImpl.createJob(BulkAPIImpl.java:37)
	at com.springml.spark.salesforce.SFObjectWriter.writeData(SFObjectWriter.scala:29)
	at com.springml.spark.salesforce.DefaultSource.updateSalesforceObject(DefaultSource.scala:154)
	at com.springml.spark.salesforce.DefaultSource.createRelation(DefaultSource.scala:130)
	at org.apache.spark.sql.execution.datasources.DataSource.write(DataSource.scala:472)
	at org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand.run(SaveIntoDataSourceCommand.scala:48)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:58)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:56)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.doExecute(commands.scala:74)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:117)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:117)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:138)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:135)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:116)
	at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:92)
	at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:92)
	at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:610)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:233)
	at $line120.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:49)
	at $line120.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:54)
	at $line120.$read$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:56)
	at $line120.$read$$iw$$iw$$iw$$iw$$iw.<init>(<console>:58)
	at $line120.$read$$iw$$iw$$iw$$iw.<init>(<console>:60)
	at $line120.$read$$iw$$iw$$iw.<init>(<console>:62)
	at $line120.$read$$iw$$iw.<init>(<console>:64)
	at $line120.$read$$iw.<init>(<console>:66)
	at $line120.$read.<init>(<console>:68)
	at $line120.$read$.<init>(<console>:72)
	at $line120.$read$.<clinit>(<console>)
	at $line120.$eval$.$print$lzycompute(<console>:7)
	at $line120.$eval$.$print(<console>:6)
	at $line120.$eval.$print(<console>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.call(IMain.scala:786)
	at scala.tools.nsc.interpreter.IMain$Request.loadAndRun(IMain.scala:1047)
	at scala.tools.nsc.interpreter.IMain$WrappedRequest$$anonfun$loadAndRunReq$1.apply(IMain.scala:638)
	at scala.tools.nsc.interpreter.IMain$WrappedRequest$$anonfun$loadAndRunReq$1.apply(IMain.scala:637)
	at scala.reflect.internal.util.ScalaClassLoader$class.asContext(ScalaClassLoader.scala:31)
	at scala.reflect.internal.util.AbstractFileClassLoader.asContext(AbstractFileClassLoader.scala:19)
	at scala.tools.nsc.interpreter.IMain$WrappedRequest.loadAndRunReq(IMain.scala:637)
	at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:569)
	at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:565)
	at scala.tools.nsc.interpreter.ILoop.interpretStartingWith(ILoop.scala:807)
	at scala.tools.nsc.interpreter.ILoop.command(ILoop.scala:681)
	at scala.tools.nsc.interpreter.ILoop.processLine(ILoop.scala:395)
	at scala.tools.nsc.interpreter.ILoop.loop(ILoop.scala:415)
	at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply$mcZ$sp(ILoop.scala:923)
	at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply(ILoop.scala:909)
	at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply(ILoop.scala:909)
	at scala.reflect.internal.util.ScalaClassLoader$.savingContextLoader(ScalaClassLoader.scala:97)
	at scala.tools.nsc.interpreter.ILoop.process(ILoop.scala:909)
	at org.apache.spark.repl.Main$.doMain(Main.scala:70)
	at org.apache.spark.repl.Main$.main(Main.scala:53)
	at org.apache.spark.repl.Main.main(Main.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:755)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:205)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:119)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
java.lang.Exception: [LoginFault [ApiFault  exceptionCode='INVALID_LOGIN'
 exceptionMessage='Invalid username, password, security token; or user locked out.'
 extendedErrorDetails='{[0]}'
]
]

  at com.springml.salesforce.wave.util.SFConfig.createPartnerConnection(SFConfig.java:110)
  at com.springml.salesforce.wave.util.SFConfig.getPartnerConnection(SFConfig.java:90)
  at com.springml.salesforce.wave.impl.BulkAPIImpl.createJob(BulkAPIImpl.java:41)
  at com.springml.salesforce.wave.impl.BulkAPIImpl.createJob(BulkAPIImpl.java:37)
  at com.springml.spark.salesforce.SFObjectWriter.writeData(SFObjectWriter.scala:29)
  at com.springml.spark.salesforce.DefaultSource.updateSalesforceObject(DefaultSource.scala:154)
  at com.springml.spark.salesforce.DefaultSource.createRelation(DefaultSource.scala:130)
  at org.apache.spark.sql.execution.datasources.DataSource.write(DataSource.scala:472)
  at org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand.run(SaveIntoDataSourceCommand.scala:48)
  at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:58)
  at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:56)
  at org.apache.spark.sql.execution.command.ExecutedCommandExec.doExecute(commands.scala:74)
  at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:117)
  at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:117)
  at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:138)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:135)
  at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:116)
  at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:92)
  at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:92)
  at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:610)
  at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:233)
  ... 48 elided
Caused by: com.sforce.soap.partner.fault.LoginFault
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  at java.lang.Class.newInstance(Class.java:442)
  at com.sforce.ws.bind.TypeMapper.readSingle(TypeMapper.java:687)
  at com.sforce.ws.bind.TypeMapper.readObject(TypeMapper.java:562)
  at com.sforce.ws.transport.SoapConnection.parseDetail(SoapConnection.java:236)
  at com.sforce.ws.transport.SoapConnection.createException(SoapConnection.java:210)
  at com.sforce.ws.transport.SoapConnection.receive(SoapConnection.java:156)
  at com.sforce.ws.transport.SoapConnection.send(SoapConnection.java:99)
  at com.sforce.soap.partner.PartnerConnection.login(PartnerConnection.java:816)
  at com.sforce.soap.partner.PartnerConnection.<init>(PartnerConnection.java:426)
  at com.sforce.soap.partner.Connector.newConnection(Connector.java:27)
  at com.springml.salesforce.wave.util.SFConfig.createPartnerConnection(SFConfig.java:107)
  ... 69 more
```

**Comment by samuel-pt** [comment]
@drake-yesware This seems to be issue with your login credentials. I just checked the same over here and it is working fine.

If still the issue exists, can you try querying salesforce using below snippet

```
val soql = "select id, name, amount from opportunity"

val sfDF = spark.
                read.
                format("com.springml.spark.salesforce").
                option("login", "https://test.salesforce.com").
                option("username", "your_salesforce_username").
                option("password", "your_salesforce_password_with_secutiry_token"). //<salesforce login password><security token>
                option("soql", soql).
                option("version", "37.0").
                load()

```

And also try to use the latest package which is `com.springml:spark-salesforce_2.11:1.1.1`
Just curious, are you trying these from spark-shell?

**Comment by drake-yesware** [comment]
Yes, this is all on spark-shell. Ok, I tried running the above soql code using `spark-shell --packages com.springml:spark-salesforce_2.11:1.1.1` on Spark 2.2.0 / Scala 2.11.8. I get the same `Invalid username, password, security token; or user locked out.` In another tab, I'm able to connect just fine using the same credentials with a Node.JS library.

**Comment by samuel-pt** [comment]
Are there any special characters in your password or securityToken? If yes, can you try escaping it?

**Comment by drake-yesware** [comment]
There is a `!` character. I tried using triple quotes around the whole <password><token> string, but still no luck. 

I'm still able to run that query successfully if I create a connection directly from `com.sforce.soap.partner.PartnerConnection` and run it with the query method though.

**Comment by drake-yesware** [comment]
I created a new salesforce account with a `-` instead of a `!` character in the password. That did not work either. 

I also enabled logging, these are the log entries right before the error:

```
18/05/04 16:13:52 INFO executor.Executor: Finished task 2.0 in stage 3.0 (TID 14). 1072 bytes result sent to driver
18/05/04 16:13:52 INFO executor.Executor: Finished task 0.0 in stage 3.0 (TID 12). 1115 bytes result sent to driver
18/05/04 16:13:52 DEBUG scheduler.TaskSchedulerImpl: parentName: , name: TaskSet_3.0, runningTasks: 1
18/05/04 16:13:52 DEBUG scheduler.TaskSchedulerImpl: parentName: , name: TaskSet_3.0, runningTasks: 0
18/05/04 16:13:52 INFO scheduler.TaskSetManager: Finished task 2.0 in stage 3.0 (TID 14) in 41 ms on localhost (executor driver) (3/4)
18/05/04 16:13:52 INFO scheduler.TaskSetManager: Finished task 0.0 in stage 3.0 (TID 12) in 43 ms on localhost (executor driver) (4/4)
18/05/04 16:13:52 INFO scheduler.TaskSchedulerImpl: Removed TaskSet 3.0, whose tasks have all completed, from pool
18/05/04 16:13:52 INFO scheduler.DAGScheduler: ResultStage 3 (collect at Utils.scala:96) finished in 0.045 s
18/05/04 16:13:52 DEBUG scheduler.DAGScheduler: After removal of stage 3, remaining stages = 0
18/05/04 16:13:52 INFO scheduler.DAGScheduler: Job 3 finished: collect at Utils.scala:96, took 0.061311 s
18/05/04 16:13:53 INFO salesforce.DefaultSource: no of partitions after repartitioning is 1
18/05/04 16:13:53 INFO salesforce.DefaultSource: Writing data
18/05/04 16:13:53 DEBUG util.ClosureCleaner: +++ Cleaning closure <function1> (com.springml.spark.salesforce.SFObjectWriter$$anonfun$1) +++
18/05/04 16:13:53 DEBUG util.ClosureCleaner:  + declared fields: 1
18/05/04 16:13:53 DEBUG util.ClosureCleaner:      public static final long com.springml.spark.salesforce.SFObjectWriter$$anonfun$1.serialVersionUID
18/05/04 16:13:53 DEBUG util.ClosureCleaner:  + declared methods: 2
18/05/04 16:13:53 DEBUG util.ClosureCleaner:      public final java.lang.Object com.springml.spark.salesforce.SFObjectWriter$$anonfun$1.apply(java.lang.Object)
18/05/04 16:13:53 DEBUG util.ClosureCleaner:      public final java.lang.String com.springml.spark.salesforce.SFObjectWriter$$anonfun$1.apply(org.apache.spark.sql.Row)
18/05/04 16:13:53 DEBUG util.ClosureCleaner:  + inner classes: 1
18/05/04 16:13:53 DEBUG util.ClosureCleaner:      com.springml.spark.salesforce.SFObjectWriter$$anonfun$1$$anonfun$apply$1
18/05/04 16:13:53 DEBUG util.ClosureCleaner:  + outer classes: 0
18/05/04 16:13:53 DEBUG util.ClosureCleaner:  + outer objects: 0
18/05/04 16:13:53 DEBUG util.ClosureCleaner:  + populating accessed fields because this is the starting closure
18/05/04 16:13:53 DEBUG util.ClosureCleaner:  + fields accessed by starting closure: 0
18/05/04 16:13:53 DEBUG util.ClosureCleaner:  + there are no enclosing objects!
18/05/04 16:13:53 DEBUG util.ClosureCleaner:  +++ closure <function1> (com.springml.spark.salesforce.SFObjectWriter$$anonfun$1) is now cleaned +++
18/05/04 16:13:53 WARN salesforce.SFObjectWriter: SaveMode ErrorIfExists Not supported. Using 'insert' operation
18/05/04 16:13:53 DEBUG util.SFConfig: Connecting SF Partner Connection using <correct email>
18/05/04 16:13:53 INFO util.SFConfig: loginURL : https://test.salesforce.com/services/Soap/u/35.0
18/05/04 16:13:53 ERROR util.SFConfig: Exception while creating connection
[LoginFault [ApiFault  exceptionCode='INVALID_LOGIN'
 exceptionMessage='Invalid username, password, security token; or user locked out.'
 extendedErrorDetails='{[0]}'
]
```

It gets to this function here: https://github.com/springml/salesforce-wave-api/blob/60dc6799d75352d7db3533730b7759fc85ffb7fe/src/main/java/com/springml/salesforce/wave/util/SFConfig.java#L96

Is there any way to confirm the password is getting passed to that function?

**Comment by samuel-pt** [comment]
No, there is no way to print the password being sent to Salesforce.

May be you can add print statement and rebuild it to see what is being sent to Salesforce.

Also, can you double check whether SOAP API access is enabled in your Salesforce account?

**Comment by drake-yesware** [comment]
Ok, I got this to connect now - though I'll still not much closer to actually understanding what happened. I set up a local server to catch the POST from the login request and turns out the password that was getting sent to the server was a 20 character string that had nothing to do with my password. Its not a base64 version of my password / token either. AFAIK its a nonrandom string value and I don't know where it came from. I get the same value across multiple shell sessions, across multiple passwords. 

I've been running Spark-shell in a docker container with a couple file volumes mounted. So far, I can only get the password to come through when I run the container without any volumes attached. Which is odd, but there must be something happening there. 

At any rate, I don't think its the library causing this. Thanks for the help @samuel-pt @springml

**Comment by springml** [comment]
We have tested this on a sandbox using "https://test.salesforce.com" as the login url.  And it worked fine.  Not exactly sure what's going on with your code.  Can you paste your code here?

**Comment by drake-yesware** [comment]
@springml I found out why this wasn't working. I had set a bunch of ENV variables. One of those sets `SF_PASSWORD`, which is unrelated to salesforce. Under some circumstances, the value from `SF_PASSWORD` env variable overrides the value of the password that is explicitly passed into spark-salesforce. 

I wasn't expecting this behavior. I must have missed it in https://github.com/springml/spark-salesforce/blob/d860d954d4ef13d58e029e6bac4cd5bb8d36a4f5/src/main/scala/com/springml/spark/salesforce/DefaultSource.scala#L302 I've solved it by removing that variable from the context my script runs in. We can close this.
